### PR TITLE
Include side-effects into function summaries.

### LINKED ISSUE #31: Summarize function bodies
Implement the code to extract a suitable summary from the state constructed by the visitor for a function body.

### PR DESCRIPTION
**Pull request by hermanventer**
## Description

This is state tracking parts of constructing and consuming function summaries.

Fixes #31 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] API change with a documentation update
- [ ] Additional test coverage

## How Has This Been Tested?

New test cases.

### CODE REVIEW

File: src/abstract_value.rs [hunk 1]
```diff
@@ -236,6 +236,14 @@ impl AbstractValue {
         }
     }
 
+    /// Returns a value that is simplified (refined) by using replacing parameter values
```
Comment (zlandau) [inline src/abstract_value.rs]: using replaced?

File: src/summaries.rs [hunk 2]
```diff
@@ -85,15 +85,18 @@ pub struct Summary {
...
     preconditions: &List<AbstractValue>,
     post_conditions: &List<AbstractValue>,
     unwind_condition: &List<AbstractValue>,
+    unwind_environment: &Environment,
 ) -> Summary {
-    let result = environment.value_at(&Path::LocalVariable { ordinal: 0 });
-    let side_effects: List<(Path, AbstractValue)> = List::new(); // todo: #31  extract from environment
-    let unwind_side_effects: List<(Path, AbstractValue)> = List::new(); // todo: #31  extract from environment
+    let result = exit_environment.value_at(&Path::LocalVariable { ordinal: 0 });
+    let side_effects: List<(Path, AbstractValue)> =
```
Comment (zlandau) [inline src/summaries.rs]: I guess I missed this before, but why Lists instead of Vecs?

File: src/abstract_value.rs (hunk 1 above)
Comment (hermanventer) [reply src/abstract_value.rs]: Fixed in a later PR.

File: src/summaries.rs (hunk 2 above)
Comment (hermanventer) [reply src/summaries.rs]: Initially I thought that these things would be part of the per block environment, so having a functional data structure would increase sharing. As things turned out, however, a Vec will do nicely instead. I'll change it.

**Pull request merged**